src/python/pants/backend/python/subsystems/python_tool_base.py

```diff
@@ -47,6 +53,43 @@ class PythonToolRequirementsBase(Subsystem):
     default_lockfile_resource: ClassVar[tuple[str, str] | None] = None
     default_lockfile_url: ClassVar[str | None] = None
 
+    install_from_resolve = StrOption(
+        advanced=True,
+        default=None,
+        help=lambda cls: softwrap(
+            f"""\
+            If specified, install the tool using the lockfile for this named resolve.
+
+            This resolve must be defined in [python].resolves, as described in
+            {doc_url("python-third-party-dependencies#user-lockfiles")}, and its lockfile must
+            provide the requirements named in the `package_names` option.
+
+            If unspecified, and the `lockfile` option is unset, the tool will be installed
+            using the default lockfile shipped with Pants.
+
+            If unspecified, and the `lockfile` option is set, the tool will use the custom
+            `{cls.options_scope}` "tool lockfile" generated from the `version` and
+            `extra_requirements` options. But note that this mechanism will soon be deprecated.
+            """
+        ),
+    )
+    # TODO: After we deprecate and remove the tool lockfile concept, we can remove the
+    #  version and extra_requirements options and directly list the default versionless
```

Comment:
Versionless is not a great idea. For most tools we actually rely on features to call them and so some loose constraint makes sense.

Reply:
Good point. I will correct this comment.

Reply:
So in that case project_names is not a good name for this option either.

Reply:
Yeah, they're requirements.